_posts/2020-07-23-demo-fraud-detection-3.md

```diff
@@ -0,0 +1,655 @@
+                              decisionTime,
+                              value.getWrapped(),
+                              aggregateResult));
+    }
+
+    // Register timers to ensure state cleanup
+    long cleanupTime = (currentEventTime / 1000) * 1000;
+    ctx.timerService().registerEventTimeTimer(cleanupTime);
+  }
+```
+
+<br/>
+Here are the details of the steps:  
+1)  We first add each new event to our window state:
+
+```java
+static <K, V> Set<V> addToStateValuesSet(MapState<K, Set<V>> mapState, K key, V value)
+      throws Exception {
+    Set<V> valuesSet = mapState.get(key);
+    if (valuesSet != null) {
+      valuesSet.add(value);
+    } else {
+      valuesSet = new HashSet<>();
+      valuesSet.add(value);
+    }
+    mapState.put(key, valuesSet);
+    return valuesSet;
+}
+```
+
+2) Next, we retrieve the previously-broadcasted rule, according to
+    which the incoming transaction needs to be evaluated.
+
+3) `getWindowStartTimestampFor` determines, given the window span defined
+    in the rule, and the current transaction timestamp, how far back in
+    time our evaluation should span.
+
+4) The aggregate value is calculated by iterating over all window state
+    entries and applying an aggregate function. It could be an *average,
+    max, min* or, as in the example rule from the beginning of this
+    section, a *sum*.
+
+```java
+private boolean isStateValueInWindow(
+    Long stateEventTime, Long windowStartForEvent, long currentEventTime) {
+  return stateEventTime >= windowStartForEvent && stateEventTime <= currentEventTime;
+}
+
+private void aggregateValuesInState(
+    Long stateEventTime, SimpleAccumulator<BigDecimal> aggregator, Rule rule) throws Exception {
+  Set<Transaction> inWindow = windowState.get(stateEventTime);
+  for (Transaction event : inWindow) {
+    BigDecimal aggregatedValue =
+        FieldsExtractor.getBigDecimalByName(rule.getAggregateFieldName(), event);
+    aggregator.add(aggregatedValue);
+  }
+}
+```
+
+5) Having an aggregate value, we can compare it to the threshold value
+    that is specified in the rule definition and fire an alert, if
+    necessary.
+
+6) At the end, we register a clean-up timer using
+`ctx.timerService().registerEventTimeTimer()`. This timer will be
+    responsible for removing the current transaction when it is going to
+    move out of scope.
+
+<div class="alert alert-info" markdown="1">
+<span class="label label-info" style="display: inline-block"><span class="glyphicon glyphicon-info-sign" aria-hidden="true"></span>  Note </span>
+Notice the rounding during timer creation. It is an important technique
+which enables a reasonable trade-off between the precision with which
+the timers will be triggered, and the number of timers being used.
+Timers are stored in Flink's fault-tolerant state, and managing them
+with millisecond-level precision can be wasteful. In our case, with this
+rounding, we will create at most one timer per key in any given second.
```

Comment:
You could link to https://ci.apache.org/projects/flink/flink-docs-release-1.11/dev/stream/operators/process_function.html#timer-coalescing if you want.